src/basic/in-addr-util.c

```diff
@@ -464,3 +464,44 @@ int in_addr_mask(int family, union in_addr_union *addr, unsigned char prefixlen)
 
         return -EAFNOSUPPORT;
 }
+
+int in_addr_prefix_from_string(const char *p, int *family, union in_addr_union *prefix, uint8_t *prefixlen) {
+        union in_addr_union buffer;
+        const char *e, *l;
+        uint8_t k;
+        int f, r;
+
+        assert(p);
+
+        /* prefixlen */
+        e = strchr(p, '/');
+        if (e) {
+                r = safe_atou8(e + 1, &k);
+                if (r < 0)
+                        return r;
+
+                l = strndupa(p, e - p);
+        } else
+                l = p;
+
+        r = in_addr_from_string_auto(l, &f, &buffer);
+        if (r < 0)
+                return r;
+
+        switch(f) {
+        case AF_INET6:
+                if (k > 128)
+                        return -ERANGE;
+                break;
+        case AF_INET:
+                if (k > 31)
+                        return -ERANGE;
+                break;
+        }
+
+        *family = f;
+        *prefix = buffer;
+        *prefixlen = k;
+
+        return 0;
```

Comment:
it would be good if we'd have a test case for this function, that checks a number of different valid and invalid inputs and verifies this works correctly.